### PR TITLE
mingw-w64 5.0.1 (new formula)

### PR DESCRIPTION
**Pull request by fxcoudert**
This is intended to be a clean PR for mingw-w64 integration into homebrew-core. I have incorporated and fixed all comments (hopefully) from the previous PR at https://github.com/Homebrew/homebrew-core/pull/9037

The layout chosen, following discussion in the previous PR, is the following:

- Inside `/usr/local/Cellar/mingw-w64/5.0.1`, we have two directories with 32- and 64-bit toolchains, named `toolchain-i686` and `toolchain-x86_64` respectively. Inside these directories, the layout is completely standard (for cross-compilers).
- In `/usr/local/Cellar/mingw-w64/5.0.1/bin`, we symlink all binutils and cross-compiler binaries, with their standard cross-compiler names, such as `x86_64-w64-mingw32-g++`, `i686-w64-mingw32-as`, etc. They get symlinked into `/usr/local/bin` as usual.
- All headers and libraries remain hidden inside our toolchains directories, so they don't conflict with native compiler files in `/usr/local/lib` and `/usr/local/include`.
-  The cross-compilers have no trouble finding their libraries (`libgcc_s.a` among others) because I removed `--enable-version-specific-runtime-libs`: since we have everything neatly in place in separate directories, no need for that option (which is apparently not that well suited for cross-compilers).
- The multilib option is commented out, but all the necessary parts are in place. I will file a GCC bug about it when I figure out exactly why it fails.
- I have tried to streamline/simplify the formula as much as possible.

I welcome all comments and reviews, and hope we can work to get this merged!

-----

- [X] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [X] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [X] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [X] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

### CODE REVIEW

**Comment by fxcoudert** [comment]
@MikeMcQuaid you had reviewed the previous PR, can you comment/review/approve?

File: Formula/mingw-w64.rb [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+  end
+
+  def install
+    bin.mkdir
+    target_archs.each do |arch|
+      # We install each arch-specific toolchain in a "toolchain-#{arch}" subdirectory
+      installdir = "#{prefix}/toolchain-#{arch}"
+      target = "#{arch}-w64-mingw32"
+
+      # Build binutils
```
Comment (MikeMcQuaid) [inline Formula/mingw-w64.rb]: Can drop this comment.

File: Formula/mingw-w64.rb [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+    url "https://ftpmirror.gnu.org/gcc/gcc-6.3.0/gcc-6.3.0.tar.bz2"
+    sha256 "f06ae7f3f790fbf0f018f6d40e844451e6bc3b7bc96e128e63b09825c1f8b29f"
+  end
+
+  def target_archs
+    ["i686", "x86_64"]
+  end
+
+  def install
+    bin.mkdir
```
Comment (MikeMcQuaid) [inline Formula/mingw-w64.rb]: What needs this?

File: Formula/mingw-w64.rb [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+    target_archs.each do |arch|
+      # We install each arch-specific toolchain in a "toolchain-#{arch}" subdirectory
+      installdir = "#{prefix}/toolchain-#{arch}"
```
Comment (MikeMcQuaid) [inline Formula/mingw-w64.rb]: `toolchain_arch_dir` or similar would be ideal and then can drop comment too

File: Formula/mingw-w64.rb [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+    sha256 "26253bf0f360ceeba1d9ab6965c57c6a48a01a8343382130d1ed47c468a3094f"
+  end
+
+  resource "gcc" do
+    url "https://ftpmirror.gnu.org/gcc/gcc-6.3.0/gcc-6.3.0.tar.bz2"
+    sha256 "f06ae7f3f790fbf0f018f6d40e844451e6bc3b7bc96e128e63b09825c1f8b29f"
+  end
+
+  def target_archs
+    ["i686", "x86_64"]
```
Comment (MikeMcQuaid) [inline Formula/mingw-w64.rb]: `.freeze`

File: Formula/mingw-w64.rb [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+          system "../configure", *args
+          system "make"
+          system "make", "install"
+        end
+      end
+
+      # Put the newly built binutils into our PATH
+      ENV.prepend_path "PATH", "#{installdir}/bin"
+
+      # Build the mingw-w64 headers
```
Comment (MikeMcQuaid) [inline Formula/mingw-w64.rb]: Can drop this comment

File: Formula/mingw-w64.rb [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+      mkdir "mingw-w64-headers/build-#{arch}" do
+        system "../configure", "--host=#{target}", "--prefix=#{installdir}/#{target}"
+        system "make"
+        system "make", "install"
+      end
+
+      # Create a mingw symlink, expected by some tools
```
Comment (MikeMcQuaid) [inline Formula/mingw-w64.rb]: e.g. which tools (add inline)

File: Formula/mingw-w64.rb [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+      Dir["#{installdir}/bin/*"].each { |f| ln_s f, bin }
+    end
+  end
+
+  test do
+    (testpath/"hello.c").write <<-EOS.undent
+      #include <stdio.h>
+      #include <windows.h>
+      int main() { puts("Hello world!"); MessageBox(NULL, TEXT("Hello GUI!"), TEXT("HelloMsg"), 0); return 0; }
+      EOS
```
Comment (MikeMcQuaid) [inline Formula/mingw-w64.rb]: Move the EOS back two characters.

File: Formula/mingw-w64.rb [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+      # Symlinks all binaries into place
+      Dir["#{installdir}/bin/*"].each { |f| ln_s f, bin }
+    end
+  end
+
+  test do
+    (testpath/"hello.c").write <<-EOS.undent
+      #include <stdio.h>
+      #include <windows.h>
+      int main() { puts("Hello world!"); MessageBox(NULL, TEXT("Hello GUI!"), TEXT("HelloMsg"), 0); return 0; }
```
Comment (MikeMcQuaid) [inline Formula/mingw-w64.rb]: Indent this line to be <=80 characters

File: Formula/mingw-w64.rb [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+      EOS
+    (testpath/"hello.cc").write <<-EOS.undent
+      #include <iostream>
+      int main() { std::cout << "Hello, world!" << std::endl; return 0; }
+      EOS
+    (testpath/"hello.f90").write <<-EOS.undent
+      program hello ; print *, "Hello, world!" ; end program hello
+      EOS
+    compiler = { ".c" => "gcc", ".cc" => "g++", ".f90" => "gfortran" }.freeze
```
Comment (MikeMcQuaid) [inline Formula/mingw-w64.rb]: I'd just hardcode the usage of this Hash; it's harder to follow as-is

File: Formula/mingw-w64.rb [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+
+    # We match the output of `objdump` in English
+    ENV["LC_ALL"] = "C"
+    target_archs.each do |arch|
+      target = "#{arch}-w64-mingw32"
+      Dir["hello{#{compiler.keys.join(",")}}"].each do |src|
+        # Compile a test program in each language
```
Comment (MikeMcQuaid) [inline Formula/mingw-w64.rb]: Can kill this comment

File: Formula/mingw-w64.rb [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+    (testpath/"hello.cc").write <<-EOS.undent
+      #include <iostream>
+      int main() { std::cout << "Hello, world!" << std::endl; return 0; }
+      EOS
+    (testpath/"hello.f90").write <<-EOS.undent
+      program hello ; print *, "Hello, world!" ; end program hello
+      EOS
+    compiler = { ".c" => "gcc", ".cc" => "g++", ".f90" => "gfortran" }.freeze
+
+    # We match the output of `objdump` in English
```
Comment (MikeMcQuaid) [inline Formula/mingw-w64.rb]: Can kill this comment

File: Formula/mingw-w64.rb [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+    ENV["LC_ALL"] = "C"
+    target_archs.each do |arch|
+      target = "#{arch}-w64-mingw32"
+      Dir["hello{#{compiler.keys.join(",")}}"].each do |src|
+        # Compile a test program in each language
+        system "#{bin}/#{target}-#{compiler[File.extname(src)]}", "-o", "test.exe", src
+
+        # Check the format of the executable produced
```
Comment (MikeMcQuaid) [inline Formula/mingw-w64.rb]: Can kill this comment

File: Formula/mingw-w64.rb [hunk 13]
```diff
@@ -0,0 +1,177 @@
+class MingwW64 < Formula
+  desc "Minimalist GNU for Windows and GCC cross-compilers"
+  homepage "https://mingw-w64.org/"
+  url "https://downloads.sourceforge.net/project/mingw-w64/mingw-w64/mingw-w64-release/mingw-w64-v5.0.1.tar.bz2"
+  sha256 "9bb5cd7df78817377841a63555e73596dc0af4acbb71b09bd48de7cf24aeadd2"
+
+  # Multilib does not work currently
+  # option "without-multilib", "Build 64-bit compiler with multilib support"
```
Comment (MikeMcQuaid) [inline Formula/mingw-w64.rb]: Probably worth removing this comment/option and all references to it until it's working.

**Review by MikeMcQuaid — CHANGES_REQUESTED** [review]

**Comment by MikeMcQuaid** [comment]
A bunch of comments but this seems almost ready to 🚢, nice work.

File: Formula/mingw-w64.rb (hunk 2 above)
Comment (fxcoudert) [inline Formula/mingw-w64.rb]: Needed so we can symlink binaries from both toolchains later on (at the end of the loop). I moved it up so we only mkdir the directory once.

Comment (MikeMcQuaid) [inline Formula/mingw-w64.rb]: It's a no-op if done multiple times so probably worth moving it down just to be clearer as to why

Comment (fxcoudert) [inline Formula/mingw-w64.rb]: Well, apparently not such as no-op… `Errno::EEXIST: File exists - /usr/local/Cellar/mingw-w64/5.0.1/bin`

I am retesting with `mkdir_p` which should fix that.

Comment (MikeMcQuaid) [inline Formula/mingw-w64.rb]: Use `bin.mkpath`

**Comment by fxcoudert** [comment]
@MikeMcQuaid I've made all the changes (except I used `mkdir_p bin` instead of `bin.mkpath` as I didn't know the later existed… I hope it's minor enough, I can't rebuild and test right now)

**Review by MikeMcQuaid — APPROVED** [review]

**Comment by MikeMcQuaid** [comment]
Thanks for your contribution to Homebrew! Without people like you submitting PRs we couldn't run this project. You rock!

**Pull request closed without merging**